src/ca/publication.rs

```diff
@@ -0,0 +1,1671 @@
+
+// Constants for the RFC 8183 XML
+const VERSION: &str = "4";
+const NS: &[u8] = b"http://www.hactrn.net/uris/rpki/publication-spec/";
+
+const MSG: &[u8] = b"msg";
+const LIST: &[u8] = b"list";
+const SUCCESS: &[u8] = b"success";
+const PUBLISH: &[u8] = b"publish";
+const WITHDRAW: &[u8] = b"withdraw";
+const REPORT_ERROR: &[u8] = b"report_error";
+const ERROR_TEXT: &[u8] = b"error_text";
+const FAILED_PDU: &[u8] = b"failed_pdu";
+
+
+//------------ PublicationCms ------------------------------------------------
+
+// This type represents a created, or parsed, RFC 8181 CMS object.
+#[derive(Clone, Debug)]
+pub struct PublicationCms {
+    signed_msg: SignedMessage,
+    message: Message,
+}
+
+impl PublicationCms {
+    /// Creates a publication CMS for the given content and signing (ID) key.
+    /// This will use a validity time of five minutes before and after 'now'
+    /// in order to allow for some NTP drift as well as processing delay
+    /// between generating this CMS, sending it, and letting the receiver
+    /// validate it.
+    pub fn create<S: Signer>(
+        message: Message,
+        issuing_key_id: &S::KeyId,
+        signer: &S,
+    ) -> Result<Self, SigningError<S::Error>> {
+        let data = message.to_xml_bytes();
+        let validity = Validity::new(
+            Time::five_minutes_ago(),
+            Time::five_minutes_from_now()
+        );
+
+        let signed_msg = SignedMessage::create(
+            data,
+            validity,
+            issuing_key_id,
+            signer
+        )?;
+
+        Ok(PublicationCms { signed_msg, message})
+    }
+
+    /// Unpack into its SignedMessage and Message
+    pub fn unpack(self) -> (SignedMessage, Message) {
+        (self.signed_msg, self.message)
+    }
+
+    pub fn into_message(self) -> Message {
+        self.message
+    }
+
+    /// Encode this to Bytes
+    pub fn to_bytes(&self) -> Bytes {
+        self.signed_msg.to_captured().into_bytes()
+    }
+
+    /// Decodes the CMS and enclosed publication Message from the source.
+    pub fn decode(
+        bytes: &[u8]
+    ) -> Result<Self, Error> {
+        let signed_msg = SignedMessage::decode(bytes, false)
+            .map_err(|e| Error::CmsDecode(e.to_string()))?;
+
+        let content = signed_msg.content().to_bytes();
+        let message = Message::decode(content.as_ref())?;
+
+        Ok(PublicationCms { signed_msg, message })
+    }
+
+    pub fn validate(&self, issuer: &IdCert) -> Result<(), Error> {
+        self.signed_msg.validate(issuer).map_err(|e| e.into())
+    }
+
+    pub fn validate_at(
+        &self, issuer: &IdCert, when: Time
+    ) -> Result<(), Error> {
+        self.signed_msg.validate_at(issuer, when).map_err(|e| e.into())
+    }
+}
+
+
+//------------ Message -------------------------------------------------------
+
+/// This type represents all Publication Messages defined in RFC8181
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Message {
+    QueryMessage(QueryMessage),
```

Comment:
`Message::QueryMessage` looks a bit odd. Perhaps the variants should just be `Message::Query` and `Message::Reply`?

Reply:
pushed, I also fixed some other cases like `Reply::ListReply` -> `Reply::List`.